File: php-phpunit/src/WordCounter.php
<?php
namespace Wordcount;

class FileNotFoundException extends \Exception {}

/**
 * Counts words of a text and provides basic analysis of that.
 */
class WordCounter {

    private $words;

    static function fromFile($fileName) {
        if (!file_exists($fileName)) {
            throw new FileNotFoundException($fileName);
        }
        $contents = file_get_contents($fileName);
        return new WordCounter($contents);
    }

    function __construct($sentence) {
        $this->words =  preg_split("/\s+/", $sentence);
    }

    function numberOfWords() {
        sleep(0.1); // here some hard core calculation...
        return count($this->words);
    }

    /**
     * @return unique words sorted alphabetically.
     */
    function uniqueWords() {
        $uniqueWords = array_unique($this->words);
        sort($uniqueWords);
        return $uniqueWords;
    }

    function containsWord($word) {
        return in_array($word, $this->words);
    }

    function countOf($word) {
        $sum = 0;
        foreach ($this->words as $s) {
            if ($word == $s) {
                $sum++;
            }
        }

        if ($sum > 0) {
            return $sum;
        }
        return null;
    }

    /**
     * @return ratio of this word's occurrence against all words.
     */
    function ratioOf($word) {
        $count = $this->countOf($word);
        if (!$count) {
            throw new \InvalidArgumentException($word . " not in sentence");
        }
        return 1.0 * $count / $this->numberOfWords();
    }

    function firstUniqueWord() {
        return $this->uniqueWords()[0];
    }

}

File: csharp-xunit/WordCount/WordCounter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.IO;
using System.Text.RegularExpressions;

namespace Org.Codecop.WordCount
{
    /// <summary>Counts words of a text and provides basic analysis of that.</summary>
    public class WordCounter
    {
        private readonly IList<string> words;

        public WordCounter(string sentence)
        {
            words = sentence.Split("\\s+");
        }

        /// <summary>Load words from a text file.</summary>
        public WordCounter(FileInfo wordFile)
            : this(File.ReadAllText(wordFile.FullName))
        {
        }

        public int NumberOfWords()
        {
            // here some hard core calculation...
            return words.Count;
        }

        /// <returns>unique words sorted alphabetically.</returns>
        public string[] UniqueWords()
        {
            var uniqueWords = words.Distinct();
            return uniqueWords.OrderBy(w => w).ToArray();
        }

        public ICollection<string> UniqueWordsAsSequence()
        {
            return UniqueWords();
        }

        public bool ContainsWord(string word)
        {
            return words.Contains(word);
        }

        public int? CountOf(string word)
        {
            var sum = 0;
            foreach (string w in words)
            {
                if (word.Equals(w))
                {
                    sum++;
                }
            }
            if (sum > 0)
            {
                return sum;
            }
            return null;
        }

        /// <returns>ratio of this word's occurrence against all words.</returns>
        public double RatioOf(string word)
        {
            var count = CountOf(word);
            if (count == null)
            {
                throw new ArgumentException(word + " not in sentence");
            }
            return 1.0 * count.Value / NumberOfWords();
        }

        public Summary SummaryOf(string word)
        {
            return new Summary(word, CountOf(word).Value);
        }
    }

    public class Summary
    {
        public readonly string word;
        public readonly int count;

        public Summary(String word, int count)
        {
            this.word = word;
            this.count = count;
        }
    }

}
